rpcs3/Emu/Memory/vm.cpp

```diff
@@ -73,43 +76,72 @@ namespace vm
 		}
 	}
 
-	bool passive_lock(cpu_thread& cpu, bool wait)
+	static atomic_t<u64>* _register_cond_lock(const u64 lock_info)
+	{
+		while (true)
+		{
+			for (auto& lock : g_cond_locks)
+			{
+				if (!lock && lock.compare_and_swap_test(0, lock_info))
+				{
+					return &lock;
+				}		
+			}
+		}
+	}
+
+	void passive_lock(cpu_thread& cpu)
 	{
 		if (UNLIKELY(g_tls_locked && *g_tls_locked == &cpu))
 		{
-			return true;
+			return;
 		}
 
 		if (LIKELY(g_mutex.is_lockable()))
 		{
 			// Optimistic path (hope that mutex is not exclusively locked)
 			_register_lock(&cpu);
 
-			if (UNLIKELY(!g_mutex.is_lockable()))
+			if (LIKELY(g_mutex.is_lockable()))
 			{
-				passive_unlock(cpu);
-
-				if (!wait)
-				{
-					return false;
-				}
-
-				::reader_lock lock(g_mutex);
-				_register_lock(&cpu);
+				return;
 			}
+
+			passive_unlock(cpu);
 		}
-		else
+
+		::reader_lock lock(g_mutex);
+		_register_lock(&cpu);
+	}
+
+	atomic_t<u64>* passive_lock(const u32 addr, const u32 size)
+	{
+		static const auto test_addr = [](const u32 target, const u32 addr, const u32 size)
 		{
-			if (!wait)
+			return addr > target || (addr + size) <= target;
```

Comment:
Condition seems wrong.

Reply:
I see, you are feeding it aligned addr / size. But function interface is at least misleading.

Reply:
its not wrong:
- target represents the starting address of the writer lock target. 
- keep in mind addr is aligned down to 128 bytes and size is aligned up:
1) if addr is bigger than target, then it cant be on the same cache line.
2) if addr + size (ie end of transfer) equal or lower than target, than it also cant be on the same cache line